### PR TITLE
Add support for pre-commit

### PR DESCRIPTION
**Pull request by loganswartz**
This allows you to use Vermin as a [pre-commit](https://pre-commit.com/) hook. The only setup required (outside of the normal pre-commit setup steps) is setting a target version for Vermin, which can either be set in a Vermin config file in your package, or directly via the `args` option in the `.pre-commit-config.yaml` (example shown in the README).

The `rev` value in the example hook configuration in the README should be set to the next git tag (likely `v1.3.4`?) when that release is made, and updated whenever a new version is released.

### CODE REVIEW

**Comment by netromdk** [comment]
Thanks for the PR, @loganswartz! I didn't know about [pre-commit](https://pre-commit.com/).

This PR is a little off because it will run vermin sequentially on each file staged for commit (or all files sequentially via `pre-commit run --all-files`). That isn't the correct way because it'll give a result for each file and that will, in almost all cases, differ from the final result of analyzing all files with one vermin invocation.

If you can setup the hook so that it will only run vermin _once_ per `git commit` invocation but with the root of the repository as input, then it should work as intended.

**Comment by loganswartz** [comment]
Good catch. According to the first line of [this](https://pre-commit.com/#arguments-pattern-in-hooks) section of their docs, they make it seem like pre-commit should pass all the staged files as a list of arguments, using only a single invocation of the hook. However, after testing, it looks like pre-commit will try to intelligently split up the workload if you've changed enough files (in my testing, it called each process with 7 files). I just added the `require_serial` option, which forces it to only call the hook once with all the files.

**Comment by netromdk** [comment]
This change fixes the case of using `pre-commit run --all-files`.

However, it still doesn't fix the real issue: that it runs vermin on the files staged and not all files. Let's say the minimum version required for a project is py3.5 but the file being staged/changed has a lower minimum version requirement, like py3.2 then vermin will report an error. It's perfectly fine to have a lower minimum requirement on a file-by-file basis.

Maybe by using [`pass_filenames`](https://pre-commit.com/#hooks-pass_filenames) as `false` and then somehow forcing it to only be passed the root of the repository, then it should work. Maybe setting [`args`](https://pre-commit.com/#hooks-args) as `['.']`. The Git pre-commit hook is always passed the Git repo as CWD.

**Comment by loganswartz** [comment]
The docs encourage you to run pre-commit with `--all-files` once when setting up new hooks to ensure everything in the repo passes the test. After you've done that, any staged changes would be the only thing that might change their minimum required versions. If it's okay for certain files to have lower requirements, then why not specify a target of ex. `3.8-` instead of `3.8`?

|                                            | Pros                                                                                                                               |
|-------------------------------------|-----------------------------------------------------------------------------------------------------------------|
|Only staged files and `3.x-` | Error message is more concise<br>Less load on your machine / faster testing times<br>Your minimum needed version can actually be lower than your target (requirement <= target) |
|The whole project and `3.x` | Version targets are slightly more in line with cmdline usage<br>You know for certain that your minimum needed version isn't lower than the one you are targeting (requirement == target) |

The only tangible difference in results would be that the second approach requires there be at least one file that pulls the minimum of the project UP to 3.8. In usage, it would require you to use `3.x-` instead of `3.x`, because otherwise yeah, you'd have issues with not being able to commit when you're only working on files that need `<3.x`.

I think the easiest solution is to just break it up into 2 hooks: one that runs on everything and one that only runs on staged files. That way people can choose what behavior they want.

File: README.rst [hunk 1]
```diff
@@ -75,6 +75,22 @@ be used to check that the minimum required versions didn't change. The following
...
+
+.. code-block:: yaml
+
+  repos:
+    - repo: https://github.com/netromdk/vermin
+      rev: e88bda9
+      hooks:
+        - id: vermin
+          # specify your target version here, OR in a Vermin config file as usual:
+          args: ['-t=3.8']
```
Comment (netromdk) [inline README.rst:88]: Let's update this line to be the following in relation to the comments to the YAML file:
```suggestion
          args: ['-t=3.8-', '--violations']
```

File: .pre-commit-hooks.yaml [hunk 2]
```diff
@@ -0,0 +1,17 @@
+- id: vermin
+  name: Check minimum Python version
+  description: Checks the minimum needed Python version to run your package. Only runs on staged files (your target must be `x.y-` rather than `x.y` for it to work properly).
```
Comment (netromdk) [inline .pre-commit-hooks.yaml:3]: This is great. In addition to mentioning that it's best to use `x.y-`, I think we should also mention that `--violations` is a good idea to use here since it works well when having only one target and it shows the violating values:
```
  --violations
        Show only results that violate versions described by --target arguments,
        which are required to be specified. Verbosity mode is automatically set to
        at least 2 in order to show violations in output text, but can be increased
        if necessary.

        If no rules are triggered while used in conjunction with --target, an exit
        code 0 will still be yielded due to inconclusivity.
```

An example of the above could be a file that doesn't do much and thus results in `~2 ~3` minimums, which means it cannot conclude anything concretely. But two targets of `~2` and `~3` differ from the input of `x.y-`. So that's a good idea to include.

**Review by netromdk — COMMENTED** [review]
Really great changes! With the following incorporated, it works very well for me.

File: README.rst [hunk 3]
```diff
@@ -75,6 +75,22 @@ be used to check that the minimum required versions didn't change. The following
   script:
   - vermin -t=2.7 -t=3 project_package otherfile.py
 
+Vermin can also be used as a `pre-commit <https://pre-commit.com/>`__ hook:
+
+.. code-block:: yaml
+
+  repos:
+    - repo: https://github.com/netromdk/vermin
+      rev: e88bda9
```
Comment (netromdk) [inline README.rst:84]: Maybe we can put a placeholder value here instead? Like
```suggestion
      rev: INPUT-SHA-OR-TAG
```

File: .pre-commit-hooks.yaml (hunk 2 above)
Comment (loganswartz) [reply .pre-commit-hooks.yaml:3]: Oh, that's super handy, I didn't actually know about this argument!

File: README.rst [hunk 4]
```diff
@@ -75,6 +75,31 @@ be used to check that the minimum required versions didn't change. The following
...
+          # specify your target version here, OR in a Vermin config file as usual:
+          args: ['-t=3.8-', '--violations']
+          # (if your target is specified in a Vermin config, you may omit the `args` entry entirely)
+
+When using the hook, a target version must be specified via a Vermin config file in your package,
+or via the ``args`` option in your ``.pre-commit-config.yaml`` config. If you're passing the target
+via ``args``, it's recommended to also include ``--violations`` (shown above).
+
+If you're using the `vermin-all` hook, you can specify any target as you usually would. However, if
+you're using the `vermin` hook, your target must be in the form of `x.y-` (as opposed to `x.y`),
```
Comment (netromdk) [inline README.rst:96]: Very small thing, but double backticks are used for code entries in RST.

**Review by netromdk — APPROVED** [review]
Thank you again!
